flocker/common/_retry.py

```diff
@@ -244,3 +250,175 @@ def should_retry(e):
+        for should_retry in should_retries:
+            result = should_retry(exc_type, value, traceback)
+            if result is not None:
+                return result
+        return None
+    return composed
+
+
+def wrap_methods_with_failure_retry(obj, should_retry=None, sleep=sleep):
+    """
+    Return a wrapper around ``obj`` which automatically re-runs method calls
+    which would otherwise have failed.
+
+    :param should_retry: A one-argument callable which accepts a three-tuple of
+        exception state and returns a ``timedelta`` or raises an exception.  If
+        the call should be retried, the ``timedelta`` gives a time to delay
+        before the retry.  If an exception is raised, no further retries are
+        attempted and the exception is propagated from the method call.
+
+    :return: An object like ``obj`` but the methods of which will be retried
+        when they raise an exception.
+    """
+    if should_retry is None:
+        should_retry = retry_some_times()
+
+    return _DecoratedInstance(
+        obj, _with_retry, should_retry=should_retry, sleep=sleep,
+    )
+
+
+def _poll_until_success_returning_result(
+    should_retry, sleep, function, args, kwargs
+):
+    """
+    Call a function until it does not raise an exception or ``should_retry``
+    says it shouldn't be tried anymore, whichever comes first.
+
+    :param should_retry: A three-argument callable which determines whether
+        further retries are attempted.  If ``None`` or a ``timedelta`` is
+        returned, another retry is attempted (immediately or after sleeping for
+        the indicated interval, respectively).  If an exception is raised,
+        further tries are not attempted and the exception is allowed to
+        propagate.
+    :param sleep: A function like ``time.sleep`` to use for the delays.
+    :param function: The function to try calling.
+    :param args: Position arguments to pass to the function.
+    :param kwargs: Keyword arguments to pass to the function.
+
+    :return: The value returned by ``function`` on the first call where it
+        returns a value instead of raising an exception.
+    """
+    saved_result = []
+
+    def pollable():
+        Message.new(
+            message_type=_TRY_RETRYING,
+        ).write()
+        try:
+            result = function(*args, **kwargs)
+        except Exception as e:
+            delay = should_retry(*exc_info())
+            Message.new(
+                message_type=_TRY_FAILURE,
+                exception=str(e),
+            ).write()
+            if delay is not None:
+                sleep(delay.total_seconds())
+            return False
+        else:
+            Message.new(
+                message_type=_TRY_SUCCESS,
+                result=result,
+            ).write()
+            saved_result.append(result)
+            return True
+
+    poll_until(pollable, repeat(0.0), sleep=sleep)
+
+    return saved_result[0]
```

Comment:
It might be best to have a voice call to talk this patch over, or perhaps even pair on it—it's very different from how I was imagining the retry system. I don't understand why the patch introduces a new way to manage the delay intervals.

It would have seemed more straightforward to me to make a synchronous equivalent to `retry_failure` (modifying the API to allow for arbitrary predicates on exceptions rather than type checks), and then to `tee` the steps iterator for each of the wrapped methods. But probably I'm missing something. Am keen to talk.